File: 3860.unique-email-groups.go
package leetcode

//
// @lc app=leetcode id=3860 lang=golang
//
// [3860] Unique Email Groups
//
// Notes
// Normalize each email by lowercasing the local/domain parts, dropping anything
// after '+', and removing dots from the local part. A Go map[string]struct{} is
// used as the set of normalized addresses. Time: O(total characters). Space:
// O(number of emails).
//
// @lc code=start

import "strings"

func UniqueEmailGroups3860(emails []string) int {
	normalized := map[string]struct{}{}
	for _, email := range emails {
		parts := strings.SplitN(email, "@", 2)
		local := strings.ToLower(parts[0])
		if plus := strings.IndexByte(local, '+'); plus != -1 {
			local = local[:plus]
		}
		local = strings.ReplaceAll(local, ".", "")
		domain := strings.ToLower(parts[1])
		normalized[local+"@"+domain] = struct{}{}
	}
	return len(normalized)
}

// @lc code=end

File: 3860.unique-email-groups.java
/*
 * @lc app=leetcode id=3860 lang=java
 *
 * [3860] Unique Email Groups
 *
 * Normalize local names by lowercasing, removing dots, and ignoring everything
 * after '+'. Normalize domains by lowercasing. Count distinct normalized
 * addresses in a HashSet.
 *
 * Time: O(total characters). Space: O(number of emails).
 */

import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

// @lc code=start
class Solution {
    public int uniqueEmailGroups(String[] emails) {
        Set<String> normalized = new HashSet<>();
        for (String email : emails) {
            String[] parts = email.split("@", 2);
            String local = parts[0].toLowerCase(Locale.ROOT);
            int plus = local.indexOf('+');
            if (plus != -1) {
                local = local.substring(0, plus);
            }
            local = local.replace(".", "");
            String domain = parts[1].toLowerCase(Locale.ROOT);
            normalized.add(local + "@" + domain);
        }
        return normalized.size();
    }
}
// @lc code=end
